Add initial inventory plugin

mancabizjak opened the pull request:
Resolves #2.

The inventory plugin adds hosts from records of the specified ServiceNow table. The users can customize which table
to query, which columns to retrieve as hostvars, and what columns should be used for `ansible_host` and inventory hostname.

### Defaults

By default, the plugin queries `cmdb_ci_server` table and uses the column `ip_address` as `ansible_host`. For inventory hostnames, the plugin uses the `name` column. It retrieves columns `name`, `host_name`, `fqdn`, `ip_address`, and adds them as host vars.

### Generating inventory groups

Inventory groups can be either auto-generated or specified explicitly (but not both):
 * **Auto-generated groups**. By default, or when using a `group_by` configuration options, groups are created automatically based on the values of the provided columns. If no columns are specified, the plugin adds hosts but doesn't create any groups (hosts are left ungrouped).
* **Explicitly specified groups**. The user specifies group names under the `groups` option, and only those groups will be created. For every group, it is possible to specify criteria for filtering which records should be considered.

### Including or excluding records

To limit the records from which Ansible hosts will be created, we support simple filtering of column values with either `includes` or `excludes`. We support specifying criteria in this manner for both user-specified groups as well as for the automatically generated ones.

Behind the scenes, the filtering logic constructs  `sysparm_query` query parameter for the ServiceNow API. Note that this isn't full-blown support for sysparm query.

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+      - Miha Dolinar (@mdolin)
+      - Tadej Borovsak (@tadeboro)
+    short_description: Inventory source for ServiceNow table records.
+    version_added: "2.10"
```
We can either omit this (because the plugin will be part of the initial release) or set it to 1.0.0.

Akasurde replied:
This is the initial commit so I would say don't add this.

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+from ..module_utils.table import TableClient
+
+
+DOCUMENTATION = r"""
```
Nitpick: the indentation of the documentation block is a mix of 2 and 4 spaces. We should pick one and stick to it. Maybe 2 is the right answer here since we already use this in all other modules?

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(("now.yaml", "now.yml")):
+                return True
+            # TODO remove?
+            self.display.vvv(
+                'Skipping due to inventory source not ending in "now.yaml" nor "now.yml"'
+            )
```
I would say we leave this is. A bit of information never hurt when debugging.

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+            self.display.vvv(
+                'Skipping due to inventory source not ending in "now.yaml" nor "now.yml"'
+            )
+            return False
```
This should be dedented one level. Right now, if parent's class verification returns `False`, we exit with `None`, which is not the worst thing in the world, but also not the best ;)

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+    def _excludes_query(self, column, excludes):
+        return "^".join("{0}!={1}".format(column, i) for i in excludes)
+
+    def _sysparm_query(self, conditions):
+        param_queries = []
+        for column, val in conditions.items():
+            if val:
+                includes = val.get("includes")
+                if includes:
+                    param_queries.append(self._includes_query(column, includes))
+                excludes = val.get("excludes")
+                if excludes:
+                    param_queries.append(self._excludes_query(column, excludes))
+        if param_queries:
+            return "^".join(i for i in param_queries)
+        return None
```
Maybe we should make this method public and provide some unit tests that will serve as a source of examples?

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+                if excludes:
+                    param_queries.append(self._excludes_query(column, excludes))
+        if param_queries:
+            return "^".join(i for i in param_queries)
```
Nitpick:

```suggestion
            return "^".join(param_queries)
```

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+        inventory_hostname = record[inventory_hostname_source]
+        if inventory_hostname:
+            host = self.inventory.add_host(inventory_hostname)
+            # TODO: what if the value is empty, do we handle this?
```
I would say we print the warning and skip the bad entry. In the future, we can then add a configuration option that will toggle the behavior between skip and abort (if anyone will be interested in this kind of functionality).

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+                self.set_hostvars(host, r)
+
+    def _merge_instance_config(self, instance_config, instance_env):
+        # Pulls the values from the environment, and if necessary, overrides
+        # with configuration provided in the inventory source.
+        instance = instance_env.copy()
+        given_keys = instance_config.keys()
+        to_override = set(instance.keys()).intersection(given_keys)
+        for option in to_override:
+            instance[option] = instance_config[option]
+        return instance
```
Maybe I am missing some context here, but if I understand the code correctly, this is just a `return dict(instance_env, **instance_config)` that skips "unknown" keys in the `instance_config` dict. In order to keep things a bit more straightforward, I would assume that the configuration file loader will not return non-declared (sub)options and that we can ignore the possibility of extra keys.

Of course, ignore my comment if my assumption about the configuration loader is off and we can get the unknown keys in the instance config.

mancabizjak replied:
Yeah, the configuration file loader does return undeclared sub-options, and if we do not sanitize user input ourselves, all the extra sub-options are passed to the HTTP `Client`, which makes its `__init__` method unhappy.

tadeboro reviewed (commented):
Things look really good thus far. I did find a few minor issues that should be easily resolvable. Thank you, @mancabizjak, for getting things this far!

Akasurde commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,443 @@
+        plugin:
+            description:
+                - The name of the ServiceNow Inventory Plugin.
+                - This should always be C(servicenow.itsm.now).
```
We can make this a constant variable. Since we are not allowing any other value.

mancabizjak replied:
There is `choices: [ servicenow.itsm.now ]` on L39 that limits the set of allowable values. And the `plugin` option must be there so that the `auto` plugin knows to load our inventory plugin, right? Is there another way to do this, what exactly did you have in mind?

mancabizjak commented:
Thank you @tadeboro and @Akasurde for the comments, I addressed your feedback in the revised version. 
Beside the fixes related to the issues you raised, I made the following changes:
* Renamed `sysparm_query`  to `sysparm_query_from_conditions` and moved it outside the `InventoryPlugin` class. I also added some docs and unit tests.
* Fixed `validate_grouping_conditions` method, since it was working only for the case when the configuration source contained `group_by`, and not for `groups`.
* To simplify the code that works on top of `instance`, `groups` and `group_by` configuration options, these now default to `{}` instead of `None`.

tadeboro commented on plugins/inventory/now.py:
```diff
@@ -0,0 +1,464 @@
+        - Windows 2003 Standard
+
+# `ansible-inventory -i inventory.now.yaml --graph` output:
+# @all:
+#  |--@HP_UX:
+#  |  |--Car-2
+#  |--@Linux_Red_Hat:
+#  |  |--DatabaseServer1
+#  |  |--DatabaseServer2
+#  |  |--lnux100
+#  |  |--lnux101
+#  |--@Solaris:
+#  |  |--MailServerUS
+#  |--@ungrouped:
```
Based on the `groups` definition above, I would expect the output to contain a single group called `non_windows_prod_servers` that contains matching records. Is this a copy-paste error maybe?

mancabizjak replied:
Yes indeed, luckily that's all it was ;) Thank you for spotting this.

tadeboro reviewed (changes requested):
I just found what seems to be a documentation error. Apart from that, I have nothing to add.

tadeboro reviewed (approved):
Thank you, @mancabizjak!
